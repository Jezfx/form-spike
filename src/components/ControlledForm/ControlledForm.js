import React, { useMemo } from "react";
import { FormProvider } from "react-hook-form";

import FollowUpField from "./FollowUpField";
import {
  hasFollowUp,
  getName,
  getFollowUp,
  formatValues,
} from "./ControlledForm.utils";

export const renderFields = (fields = []) =>
  fields.map(({ Component, ...props }, index) => (
    <div key={getName(props) || index}>
      <Component {...props} />

      {hasFollowUp(props) && (
        <FollowUpField
          key={`${getName(props)}-followUp`}
          fieldName={getName(props)}
          followUp={getFollowUp(props)}
        />
      )}
    </div>
  ));

const ControlledForm = ({ onSubmit, buttons, methods, fields }) => {
  const { handleSubmit } = methods;
  const renderFormFields = useMemo(() => renderFields(fields), [fields]);
  const renderButtons = useMemo(() => (!!buttons ? buttons() : null), [
    buttons,
  ]);

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit((values) => onSubmit(formatValues(values)))}>
        {renderFormFields}
        {renderButtons}
      </form>
    </FormProvider>
  );
};

ControlledForm.defaultProps = {
  fields: [],
};

export default ControlledForm;
