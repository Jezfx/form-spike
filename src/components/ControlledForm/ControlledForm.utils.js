import { isNotEmpty } from "../../utils";
import { get, flow } from "lodash/fp";

export const getName = get("name");
export const getFollowUp = get("followUp");
export const hasFollowUp = flow([getFollowUp, isNotEmpty]);
