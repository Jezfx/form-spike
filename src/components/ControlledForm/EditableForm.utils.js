import { isNotEmpty } from "../../utils";
import { get, flow } from "lodash/fp";

export const getKey = get("group.key");
export const getFollowUp = get("group.followUp");
export const hasFollowUp = flow([getFollowUp, isNotEmpty]);
