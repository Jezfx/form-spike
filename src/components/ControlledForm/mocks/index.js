import ControlledTextField from "../../ControlledTextField";
import ControlledRadioGroup from "../../ControlledRadioGroup";

export const defaultFields = [
  {
    Component: ControlledTextField,
    name: "firstName",
    label: "Text Field",
  },
];

export const followUpFields = [
  {
    Component: ControlledRadioGroup,
    name: "level_one",
    label: "Level one (check yes to see level 2)",
    group: {
      options: [
        {
          name: "isSameName",
          label: "Yes",
          value: "yes",
        },
        {
          name: "isDifferentName",
          label: "No",
          value: "no",
        },
      ],
    },
    followUp: {
      condition: () => true,
      fields: [
        {
          label: "Other Options...",
          Component: ControlledRadioGroup,
          name: "other_options",
          group: {
            options: [
              {
                name: "isSameName",
                label: "Follow up",
                value: "yes",
              },
            ],
          },
        },
      ],
    },
  },
];
