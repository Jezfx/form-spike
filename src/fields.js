import ControlledTextField from "./ControlledTextField";
import ControlledRadioGroup from "./ControlledRadioGroup";
import ControlledFieldArray from "./ControlledFieldArray";
import ControlledCheckBox from "./ControlledCheckBox";

export const fields = [
  {
    Component: ControlledTextField,
    name: "firstName",
    label: "Text Field",
  },
  {
    Component: ControlledCheckBox,
    name: "checkboxes",
    label: "CheckBoxes",
    group: {
      key: "checkboxes",
      followUp: {
        condition: "other",
        fields: [
          {
            label: "Other Options...",
            Component: ControlledRadioGroup,
            name: "level_two",
            group: {
              key: "level_three",
              options: [
                {
                  key: "name_same",
                  name: "isSameName",
                  label: "Yes",
                  value: "yes",
                },
                {
                  key: "name_diff",
                  name: "isDifferentName",
                  label: "No",
                  value: "no",
                },
              ],
            },
          },
        ],
      },
      options: [
        {
          key: "name_same",
          name: "checkbox_one",
          label: "check box one",
          value: "foo",
        },
        {
          key: "other",
          name: "other",
          label: "Other",
          value: "other",
        },
      ],
    },
  },
  {
    Component: ControlledFieldArray,
    name: "import_countries",
    label: "Import contries",
  },
  {
    Component: ControlledRadioGroup,
    name: "level_one",
    label: "Level one (check yes to see level 2)",
    group: {
      key: "level_one",
      followUp: {
        condition: "yes",
        fields: [
          {
            label: "Level Two (check no to see level 3)",
            Component: ControlledRadioGroup,
            name: "level_two",
            group: {
              key: "level_three",
              followUp: {
                condition: "no",
                fields: [
                  {
                    Component: ControlledTextField,
                    name: "level_three_input",
                    label: "Level Three Input",
                  },
                ],
              },
              options: [
                {
                  key: "name_same",
                  name: "isSameName",
                  label: "Yes",
                  value: "yes",
                },
                {
                  key: "name_diff",
                  name: "isDifferentName",
                  label: "No",
                  value: "no",
                },
              ],
            },
          },
        ],
      },
      options: [
        {
          key: "name_same",
          name: "isSameName",
          label: "Yes",
          value: "yes",
        },
        {
          key: "name_diff",
          name: "isDifferentName",
          label: "No",
          value: "no",
        },
      ],
    },
  },
];
