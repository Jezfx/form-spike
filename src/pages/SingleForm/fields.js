import ControlledTextField from "../../components/ControlledTextField";
import ControlledRadioGroup from "../../components/ControlledRadioGroup";
import ControlledFieldArray from "../../components/ControlledFieldArray";
import ControlledCheckBox from "../../components/ControlledCheckBox";
import ControlledDropdown from "../../components/ControlledDropdown";

const dropdownOptions = [
  { label: "Please select", disabled: true },
  { value: "GBP", label: "GBP" },
  { value: "EUR", label: "EUR" },
  { value: "USD", label: "USD" },
];

export const fields = [
  {
    Component: ControlledTextField,
    name: "firstName",
    label: "Text Field",
  },
  {
    Component: ControlledCheckBox,
    name: "checkboxes",
    label: "CheckBoxes",
    group: {
      options: [
        {
          value: "foo",
          name: "checkbox_one",
          label: "check box one",
        },
        {
          label: "Other",
          name: "other",
          value: "other",
        },
      ],
    },
    followUp: {
      condition: "other",
      fields: [
        {
          label: "Other Options...",
          Component: ControlledRadioGroup,
          name: "level_two",
          group: {
            options: [
              {
                name: "isSameName",
                label: "Yes",
                value: "yes",
              },
              {
                name: "isDifferentName",
                label: "No",
                value: "no",
              },
            ],
          },
        },
      ],
    },
  },
  {
    Component: ControlledFieldArray,
    name: "import_countries",
    label: "Import contries",
  },
  {
    Component: ControlledRadioGroup,
    name: "level_one",
    label: "Level one (check yes to see level 2)",
    group: {
      options: [
        {
          name: "isSameName",
          label: "Yes",
          value: "yes",
        },
        {
          name: "isDifferentName",
          label: "No",
          value: "no",
        },
      ],
    },
    followUp: {
      condition: "no",
      fields: [
        {
          label: "Other Options...",
          Component: ControlledRadioGroup,
          name: "other_options",
          group: {
            options: [
              {
                name: "isSameName",
                label: "Yes",
                value: "yes",
              },
              {
                name: "isDifferentName",
                label: "No",
                value: "no",
              },
            ],
          },
          followUp: {
            condition: "no",
            fields: [
              {
                label: "Other Name...",
                Component: ControlledTextField,
                name: "level_three",
              },
            ],
          },
        },
      ],
    },
  },
  {
    Component: ControlledDropdown,
    name: "uk_responsible_person_country",
    label: "Country *",
    options: dropdownOptions,
  },
];
