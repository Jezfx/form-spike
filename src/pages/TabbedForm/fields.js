import ControlledTextField from "../../components/ControlledTextField";
import ControlledDropdown from "../../components/ControlledDropdown";

const dropdownOptions = [
  { id: 1, value: "GBP", label: "GBP" },
  { id: 2, value: "EUR", label: "EUR" },
  { id: 3, value: "USD", label: "USD" },
];

export const fields = [
  {
    Component: ControlledTextField,
    name: "firstName",
    label: "Text Field",
  },
  {
    Component: ControlledDropdown,
    name: "uk_responsible_person_country",
    label: "Country *",
    options: dropdownOptions,
  },
];
