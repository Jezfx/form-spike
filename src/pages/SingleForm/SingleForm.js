import React from "react";
import { useForm } from "react-hook-form";
import { Button, Typography } from "@material-ui/core";

import singleFormValidationResolver from "./validationSchema";
import ControlledForm from "../../components/ControlledForm";
import { fields } from "./fields";

const renderButtons = () => (
  <Button type="submit" variant="contained" color="primary">
    Submit
  </Button>
);

const defaultValues = {
  checkboxes: ["foo"],
  import_countries: [
    {
      value: "spain",
    },
  ],
  firstName: "sad",
  level_one: "yes",
  level_three: "no",
  level_three_input: "asdad",
  uk_responsible_person_country: "USD",
};

const SingleForm = () => {
  const methods = useForm({
    mode: "onBlur",
    reValidateMode: "onChange",
    resolver: singleFormValidationResolver,
    defaultValues,
  });

  const onSubmit = (data) => {
    console.log(data);
  };
  return (
    <>
      <Typography variant="h6">Form Title</Typography>
      <Typography variant="body2">Contextual information</Typography>
      <ControlledForm
        methods={methods}
        onSubmit={onSubmit}
        fields={fields}
        buttons={renderButtons}
      />
    </>
  );
};

export default SingleForm;
