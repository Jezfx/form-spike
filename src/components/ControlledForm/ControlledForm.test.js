import React from "react";
import {
  wait,
  fireEvent,
  cleanup,
  render,
  screen,
} from "@testing-library/react";
import { Button } from "@material-ui/core";

import ControlledForm from "./ControlledForm";
import { fields } from "../../pages/SingleForm/fields";
import { withForm } from "../../test-utils";

describe("Controlled Form", () => {
  afterEach(cleanup);

  const props = {
    buttons: () => <Button type="submit">Submit</Button>,
    fields,
    onSubmit: jest.fn(),
  };

  const CompWithForm = withForm(ControlledForm);
  it("matches snapshot", () => {
    const { asFragment } = render(<CompWithForm {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls on submit", async () => {
    render(<CompWithForm {...props} />);

    const button = screen.getByText("Submit");
    fireEvent.click(button);

    await wait(() => {
      expect(props.onSubmit).toHaveBeenCalled();
    });
  });
});
