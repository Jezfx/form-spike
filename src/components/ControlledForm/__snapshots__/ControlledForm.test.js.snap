// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controlled Form matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
          data-shrink="false"
        >
          Text Field
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiFilledInput-input"
            name="firstName"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div>
      <fieldset
        class="MuiFormControl-root"
      >
        <legend
          class="MuiFormLabel-root"
        >
          CheckBoxes
        </legend>
        <div
          class="MuiFormGroup-root"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-4"
                  data-indeterminate="false"
                  name="checkboxes[0][0]"
                  type="checkbox"
                  value="checkboxes[0][foo]"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              check box one
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-4"
                  data-indeterminate="false"
                  name="checkboxes[1][1]"
                  type="checkbox"
                  value="checkboxes[1][other]"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Other
            </span>
          </label>
        </div>
        <p
          class="MuiFormHelperText-root"
        />
      </fieldset>
    </div>
    <div>
      <fieldset
        class="MuiFormControl-root"
      >
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <div
            aria-label="outlined primary button group"
            class="MuiButtonGroup-root"
            role="group"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Add
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </fieldset>
    </div>
    <div>
      <fieldset
        class="MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root Mui-required"
        >
          Level one (check yes to see level 2)
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk"
          >
            â€‰*
          </span>
        </label>
        <div
          class="MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-4"
                  name="isSameName"
                  type="radio"
                  value="yes"
                />
                <div
                  class="PrivateRadioButtonIcon-root-6"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-7"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Yes
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-4"
                  name="isDifferentName"
                  type="radio"
                  value="no"
                />
                <div
                  class="PrivateRadioButtonIcon-root-6"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-7"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              No
            </span>
          </label>
        </div>
        <p
          class="MuiFormHelperText-root Mui-required"
        />
      </fieldset>
    </div>
    <div>
      <div
        class="MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root"
          for="uk_responsible_person_country"
        >
          Country *
        </label>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <select
              aria-invalid="false"
              class="MuiSelect-root MuiSelect-select MuiInputBase-input MuiInput-input"
              name="uk_responsible_person_country"
            >
              <option
                disabled=""
              >
                Please select
              </option>
              <option
                value="GBP"
              >
                GBP
              </option>
              <option
                value="EUR"
              >
                EUR
              </option>
              <option
                value="USD"
              >
                USD
              </option>
            </select>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
